File: producer-api-go-rest-lambda/internal/models/event.go
package models

import (
	"encoding/json"
	"fmt"
	"strconv"
	"time"
)

// Event represents the complete event structure
type Event struct {
	EventHeader EventHeader `json:"eventHeader"`
	EventBody   EventBody   `json:"eventBody"`
}

// EventHeader contains event metadata
type EventHeader struct {
	UUID        *string    `json:"uuid,omitempty"`
	EventName   string     `json:"eventName"`
	CreatedDate *time.Time `json:"createdDate,omitempty"`
	SavedDate   *time.Time `json:"savedDate,omitempty"`
	EventType   *string    `json:"eventType,omitempty"`
}

// EventBody contains the entity updates
type EventBody struct {
	Entities []EntityUpdate `json:"entities"`
}

// EntityUpdate represents a single entity update
type EntityUpdate struct {
	EntityType        string          `json:"entityType"`
	EntityID          string          `json:"entityId"`
	UpdatedAttributes json.RawMessage `json:"updatedAttributes"`
}

// UnmarshalJSON implements custom unmarshaling for EventHeader to handle flexible date parsing
func (eh *EventHeader) UnmarshalJSON(data []byte) error {
	type Alias EventHeader
	aux := &struct {
		CreatedDate interface{} `json:"createdDate,omitempty"`
		SavedDate   interface{} `json:"savedDate,omitempty"`
		*Alias
	}{
		Alias: (*Alias)(eh),
	}

	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}

	// Parse CreatedDate
	if aux.CreatedDate != nil {
		createdDate, err := parseFlexibleDate(aux.CreatedDate)
		if err != nil {
			return fmt.Errorf("invalid createdDate: %w", err)
		}
		eh.CreatedDate = createdDate
	}

	// Parse SavedDate
	if aux.SavedDate != nil {
		savedDate, err := parseFlexibleDate(aux.SavedDate)
		if err != nil {
			return fmt.Errorf("invalid savedDate: %w", err)
		}
		eh.SavedDate = savedDate
	}

	return nil
}

// parseFlexibleDate handles both ISO 8601 strings and Unix timestamps (milliseconds)
func parseFlexibleDate(v interface{}) (*time.Time, error) {
	switch val := v.(type) {
	case string:
		// Try ISO 8601 formats first
		formats := []string{
			time.RFC3339,
			time.RFC3339Nano,
			"2006-01-02T15:04:05Z",
			"2006-01-02T15:04:05.000Z",
		}
		for _, format := range formats {
			if t, err := time.Parse(format, val); err == nil {
				return &t, nil
			}
		}
		// Try parsing as numeric string (timestamp in milliseconds)
		if ms, err := strconv.ParseInt(val, 10, 64); err == nil {
			secs := ms / 1000
			nsecs := (ms % 1000) * 1000000
			t := time.Unix(secs, nsecs)
			return &t, nil
		}
		return nil, fmt.Errorf("unable to parse date string: %s", val)
	case float64:
		// JSON numbers are parsed as float64
		ms := int64(val)
		secs := ms / 1000
		nsecs := (ms % 1000) * 1000000
		t := time.Unix(secs, nsecs)
		return &t, nil
	case int64:
		ms := val
		secs := ms / 1000
		nsecs := (ms % 1000) * 1000000
		t := time.Unix(secs, nsecs)
		return &t, nil
	default:
		return nil, fmt.Errorf("unsupported date type: %T", v)
	}
}

File: producer-api-python-rest-lambda-shared/producer_api_python_rest_lambda_shared/models/event.py
"""Event models for Producer API."""

from datetime import datetime
from typing import Any, List, Optional

from dateutil import parser
from pydantic import BaseModel, Field, field_validator


class EventHeader(BaseModel):
    """Event header containing metadata."""
    
    uuid: Optional[str] = None
    event_name: str = Field(..., alias="eventName")
    created_date: Optional[datetime] = Field(None, alias="createdDate")
    saved_date: Optional[datetime] = Field(None, alias="savedDate")
    event_type: Optional[str] = Field(None, alias="eventType")
    
    @field_validator("created_date", "saved_date", mode="before")
    @classmethod
    def parse_flexible_date(cls, v: Any) -> Optional[datetime]:
        """Parse date from ISO 8601 string or Unix timestamp (milliseconds)."""
        if v is None:
            return None
        
        if isinstance(v, datetime):
            return v
        
        if isinstance(v, str):
            # Try ISO 8601 formats first
            try:
                return parser.isoparse(v)
            except (ValueError, TypeError):
                pass
            
            # Try parsing as numeric string (timestamp in milliseconds)
            try:
                ms = int(v)
                return datetime.fromtimestamp(ms / 1000.0)
            except (ValueError, TypeError):
                pass
            
            raise ValueError(f"Unable to parse date string: {v}")
        
        if isinstance(v, (int, float)):
            # JSON numbers are parsed as int/float
            ms = int(v)
            return datetime.fromtimestamp(ms / 1000.0)
        
        raise ValueError(f"Unsupported date type: {type(v)}")
    
    class Config:
        """Pydantic config."""
        populate_by_name = True
        json_encoders = {
            datetime: lambda v: v.isoformat() if v else None
        }


class EntityUpdate(BaseModel):
    """Entity update information."""
    
    entity_type: str = Field(..., alias="entityType")
    entity_id: str = Field(..., alias="entityId")
    updated_attributes: Any = Field(..., alias="updatedAttributes")
    
    class Config:
        """Pydantic config."""
        populate_by_name = True


class EventBody(BaseModel):
    """Event body containing entity updates."""
    
    entities: List[EntityUpdate]


class Event(BaseModel):
    """Complete event structure."""
    
    event_header: EventHeader = Field(..., alias="eventHeader")
    event_body: EventBody = Field(..., alias="eventBody")
    
    class Config:
        """Pydantic config."""
        populate_by_name = True
